defold-spine/src/comp_spine_model.cpp

```diff
@@ -727,25 +734,32 @@ namespace dmSpine
             }
 
             component.m_DoRender = 1;
+
+            // Update bounding boxes
+            SpineModelBounds& bounds = world->m_BoundingBoxes[i];
+            GetSkeletonBounds(component.m_SkeletonInstance,bounds);
         }
 
         // Since we've moved the child game objects (bones), we need to sync back the transforms
         update_result.m_TransformsUpdated = num_active > 0;
         return dmGameObject::UPDATE_RESULT_OK;
     }
 
+
     static void RenderBatch(SpineModelWorld* world, dmRender::HRenderContext render_context, dmRender::RenderListEntry *buf, uint32_t* begin, uint32_t* end)
     {
         //DM_PROFILE(SpineModel, "RenderBatch");
 
-        const SpineModelComponent* first = (SpineModelComponent*) buf[*begin].m_UserData;
+        uint32_t component_index = (uint32_t)buf[*begin].m_UserData;
+        const SpineModelComponent* first = (const SpineModelComponent*) world->m_Components.m_Objects[component_index];
```

Comment:
This is using a _private_ member. See `object_pool.h`
Instead, use the helper function:
```
    static inline SpineModelComponent* GetComponentFromIndex(SpineModelWorld* world, int index)
    {
        return world->m_Components.Get(index);
    }
```


Also, this index is _not_ the same as the actual index returned from the pool when creating the object. See `CompSpineModelCreate`

Reply:
Right, that was also my concern when asked about `world.m_Components.m_Objects` above.

Let me give a second look...

Reply:
I've done a comparison between the Sprites an Spine implementation. I'm afraid i can't find many differences between them: 

### In Sprites

In comp_sprite.cpp:CompSpriteRender(), the algorithm goes through `sprite_world->m_Components.m_Objects` treats them as an array and stores the loop counter (i) in `RenderListEntry.m_UserData`.
	
https://github.com/defold/defold/blob/dev/engine/gamesys/src/gamesys/components/comp_sprite.cpp#L1165

Later, in `comp_sprite.cpp:CalcBoundingVolumes()` it will iterate `sprite_world->m_Components.m_Objects` like an array and use the index (stored in RenderListEntry.m_UserData) to access `sprite_world->m_BoundingVolumes[i]` and store there the radius.

https://github.com/defold/defold/blob/dev/engine/gamesys/src/gamesys/components/comp_sprite.cpp#L982-L993

Later, in comp_sprite.cpp:RenderListFrustumCulling() it will use this index (RenderListEntry.m_UserData) again on `sprite_world->m_BoundingVolumes` but this time to read the value.
	
https://github.com/defold/defold/blob/dev/engine/gamesys/src/gamesys/components/comp_sprite.cpp#L1076


### In Spine

In comp_spine_model.cpp:CompSpineModelRender(), the algorithm goes through `world->m_Components.m_Objects` and stores the loop counter (i) in `RenderListEntry.m_UserData`. Same think as in Sprites.

https://github.com/defold/extension-spine/blob/issue-6980_spine-frustum-culling/defold-spine/src/comp_spine_model.cpp#L893-L908
	
Later, in `comp_spine_model.cpp:CompSpineModelUpdate()` it will go through `world->m_Components.m_Objects` and use the loop counter to initialize ` world->m_BoundingBoxes[i]` at the end.

https://github.com/defold/extension-spine/blob/issue-6980_spine-frustum-culling/defold-spine/src/comp_spine_model.cpp#L690-L696  
https://github.com/defold/extension-spine/blob/issue-6980_spine-frustum-culling/defold-spine/src/comp_spine_model.cpp#L739
	
Later, in comp_spine_model.cpp:RenderListFrustumCulling()  it will use the index stored in `entry->m_UserData` to get the respective component and also the bounding volume:

https://github.com/defold/extension-spine/blob/issue-6980_spine-frustum-culling/defold-spine/src/comp_spine_model.cpp#L828-L834

Reply:
For readability in this issue (since we had an offline discussion):

The difference between Sprite and Spine, is that Sprite does all the calculations within a single function (during the Render() call).
Thus, it is not possible for the object pool to change, and therefore the code can assume certain things.

In this PR though, the code is split up between Update() and Render(), using indices directly into an internal data member. In between Update() and Render(), this object pool may change, thus invalidating these indices.


I currently ee two ways of solving this.
1) Store the pool index given by the pool's Alloc() as the index.
2) Move the bounding box calculations to the top of the Render() function (using the same idea as the Sprite component)

I think number 2 is the easiest way forward.
Note, I've created this [Sprite PR](https://github.com/defold/defold/pull/7054) where access to the internal data is restricted, and I've also added more comments in the Sprite implementation about the requirements/assumptions about data lifetime)